packages/client/lib/sync/execution/vmexecution.ts

```diff
@@ -49,6 +49,7 @@ export class VMExecution extends Execution {
     const blockNumber = headBlock.header.number.toNumber()
     this.config.execCommon.setHardforkByBlockNumber(blockNumber)
     this.hardfork = this.config.execCommon.hardfork()
+    this.vm._updateOpcodes()
```

Comment:
Side note: we can also launch VM with the `selectHardforkByBlockNumber: true` option, then this stuff is handled in the VM. We would only not directly see in Client that we are switching HFs. This would generally go well if we always start the Common with Chainstart, and then let VM check per block number (via this option) if we should change it.

Reply:
Great find: already thought some time ago: would it make sense for such uses cases to make the `Common` also a class inheriting from `EventEmitter` and then submit an event `hardforkChanged` whenever a HF change has been triggered?

Then the VM can listen to this internally and itself do a call to `_updateOpcodes()`. Calling this from the outside is not so optimal and likely other users will stumble upon this as well.

(making a class an EventEmitter should be a non-breaking change I would say?)

Reply:
Yeah, this would make a lot of sense. I think at the moment our packages are mostly built around non-changing Common hardforks. We also have this issue for instance: #969. Although I'm not sure if reacting on events would solve that particular problem in an elegant way 🤔 .

Reply:
And it would indeed be non-breaking, since you only add methods. (Unless I'm missing some very specific things which might be broken if you suddenly start extending classes).